src/NServiceBus.Core/Unicast/Messages/MessageMetadataRegistry.cs

```diff
@@ -48,12 +48,21 @@ public MessageMetadata GetMessageMetadata(string messageTypeIdentifier)
         {
             Guard.AgainstNullAndEmpty(nameof(messageTypeIdentifier), messageTypeIdentifier);
 
-            var messageType = Type.GetType(messageTypeIdentifier, false);
+            var messageType = GetType(messageTypeIdentifier);
 
             if (messageType == null)
             {
                 Logger.DebugFormat("Message type: '{0}' could not be determined by a 'Type.GetType', scanning known messages for a match", messageTypeIdentifier);
-                return messages.Values.FirstOrDefault(m => m.MessageType.FullName == messageTypeIdentifier);
+
+                foreach (var item in messages.Values)
```

Comment:
Looking at this more, I think both my new `cachedTypes` collection and the existing `messages` collection could have concurrency issues here.

The [docs ](https://msdn.microsoft.com/en-us/library/xfhwa508(v=vs.110).aspx#Anchor_10)say that it's only safe to concurrently read, and we could have both reading and writing going on. We could even have multiple writers, which I think could actually end up with lost items from the collection if both writes trigger an internal data structure reorganization.